src/FieldMapB.cpp

```diff
@@ -0,0 +1,377 @@
+}
+
+// get XYZ cell indices corresponding to point of interest
+bool FieldMapB::GetIndices(double X, double Y, double Z, int *idxX, int *idxY, int *idxZ, double *deltaX, double *deltaY, double *deltaZ)
+{
+  // boundary check
+  if( X > maxs[0] || X < mins[0] || Y > maxs[1] || Y < mins[1] || Z > maxs[2] || Z < mins[2] ) {
+    return false;
+  }
+
+  // get indices
+  *deltaX = std::modf( (X - mins[0]) / steps[0], &ix_d );
+  *deltaY = std::modf( (Y - mins[1]) / steps[1], &iy_d );
+  *deltaZ = std::modf( (Z - mins[2]) / steps[2], &iz_d );
+
+  *idxX = static_cast<int>( ix_d );
+  *idxY = static_cast<int>( iy_d );
+  *idxZ = static_cast<int>( iz_d );
+
+  return true;
+}
+
+// load data
+void FieldMapB::LoadMap(const std::string& map_file, double scale)
+{
+  std::string   line;
+  std::ifstream input(map_file);
+  if (!input) {
+    std::cout << "FieldMapB Error: file \"" << map_file << "\" cannot be read." << std::endl;
+  }
+
+  std::vector<double> coord = {};
+  std::vector<double> Bcomp = {};
+
+  while (std::getline(input, line).good()) {
+    std::istringstream iss(line);
+
+    coord.clear();
+    Bcomp.clear();
+
+    if( case_BrBz ) {
+      coord.resize(2);
+      Bcomp.resize(2);
+      iss >> coord[0] >> coord[1] >> Bcomp[0] >> Bcomp[1];
+
+      if( ! GetIndices( coord[0], coord[1], &ir, &iz, &dr, &dz) ) {
+        std::cout << "FieldMapB Warning: coordinates out of range, skipped it."<< std::endl;
+      }
+      else {
+        Bvals_RZ[ ir ][ iz ] = { Bcomp[0] * scale, Bcomp[1] * scale };
+        //ROOT::Math::XYZPoint p(coord[0], 0, coord[1]);
+        //std::cout << p << " -> " << trans*p << std::endl;
+        //std::cout << indices[0] << ", " << indices[1] << ", " << Bcomp[0] << ", " << Bcomp[1] << std::endl;
+      }
+    }
+    else {
+      coord.resize(3);
+      Bcomp.resize(3);
+      iss >> coord[0] >> coord[1] >> coord[2] >> Bcomp[0] >> Bcomp[1] >> Bcomp[2];
+
+      if( ! GetIndices(coord[0], coord[1], coord[2], &ix, &iy, &iz, &dx, &dy, &dz) ) {
+        std::cout << "FieldMapB Warning: coordinates out of range, skipped it."<< std::endl;
+      }
+      else {
+        Bvals_XYZ[ ix ][ iy ][ iz ] = { Bcomp[0] * scale, Bcomp[1] * scale, Bcomp[2] * scale };
+      }
+    }
+  }
+}
+
+// get field components
+void FieldMapB::fieldComponents(const double* pos, double* field)
+{
+  // coordinate conversion
+  auto p = trans_inv * ROOT::Math::XYZPoint(pos[0], pos[1], pos[2]);
+
+  if( case_BrBz ) {
+    // coordinates conversion
+    const double r   = sqrt(p.x() * p.x() + p.y() * p.y());
+    const double z   = p.z();
+    const double phi = atan2(p.y(), p.x());
+
+    if( ! GetIndices(r, z, &ir, &iz, &dr, &dz) ) {
+      // out of range
+      return;
+    }
+
+    // p1    p3
+    //    p
+    // p0    p2
+    auto& p0 = Bvals_RZ[ ir     ][ iz    ];
+    auto& p1 = Bvals_RZ[ ir     ][ iz + 1];
+    auto& p2 = Bvals_RZ[ ir + 1 ][ iz    ];
+    auto& p3 = Bvals_RZ[ ir + 1 ][ iz + 1];
+
+    // Bilinear interpolation
```

Comment:
As an optional exercise in a future pr, it may be fun to see how much of this can be 1) modified to use fma, fused multiply-add (single CPU op) or how much the compiler already figures that out, 2) use veccore or std:: experimental::simd types to parallelize the two or three vector components that need the same operation anyway. Or, of course, one would start with figuring out how much it actually matters...

Reply:
I've never looked this much into optimization, although I'm curious how it could be done.
So, is the point that a long operation, i.e. a*b + c*d + ..., won't fit into a single CPU op, which means it takes longer? 
What kinds of operations fit into one operation?